web/src/stream_topic_history.ts

```diff
@@ -234,9 +276,36 @@ export function remove_messages(opts: {
         return;
     }
 
+    if (opts.topic_deleted) {
+        // topic_deleted is true only when propagate_mode is
+        // set to change_all and here we can be sure that the
+        // topic will no longer exist, so we will remove the
+        // topic completely.
+        history.topics.delete(topic_name);
+        return;
+    }
+
     // Update max_message_id in topic
     if (existing_topic.message_id <= max_removed_msg_id) {
         const msgs_in_topic = message_util.get_messages_in_topic(stream_id, topic_name);
+
+        if (!msgs_in_topic.length) {
+            get_latest_message_in_topic(stream_id, topic_name, (latest_message_id) => {
+                if (!latest_message_id) {
+                    history.topics.delete(topic_name);
+                } else {
+                    existing_topic.message_id = latest_message_id;
+                    existing_topic.is_data_reliable = true;
+                }
+                if (topic_list.active_stream_id() === stream_id) {
+                    const $stream_li = topic_list.get_stream_li();
+                    assert($stream_li !== undefined);
+                    topic_list.rebuild($stream_li, stream_id);
+                }
```

Comment:
Can you add a comment hinting why this block is needed? It's not obvious to me

Reply:
I intended to mention this before but forgot. One problem here is that this leads to rebuilding of topic list twice.

Reply:
Added the comment for this block in `remove_messages` code and at two places in `add_moved_messages` code - one where we get data for only topic from the server and one where we get the data for all topics in the stream.